Issue/14055 - Mark referrer as spam

nikola-milicevic opened the pull request:
Fixes [#14055](https://github.com/wordpress-mobile/WordPress-iOS/issues/14055)

Depends on [WordPressKit PR](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/407).

### Description:

This PR adds marking / unmarking as spam functionality to the referrer.

### To test:
1. On 'My Site' tab go to 'Stats' screen.
2. Tap on Days / Weeks / Month / Years
3. Tap on a referrer to mark it as spam.
4. Refreshing the data will remove the referrer marked as spam from the list.

Note: if there are many referrers, marking referrer as spam can be done from the referrers list screen as well (accessed by tapping on 'View more' button).

### Video:



https://user-images.githubusercontent.com/12729242/122479467-6c774a80-cfcb-11eb-8dc8-2bd42508984d.mov





### Regression Notes
1. Potential unintended areas of impact
n/a

2. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a

3. What automated tests I added (or what prevented me from doing so)
- StatsPeriodStore mark / unmark referrer as spam

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

nikola-milicevic committed:
Add invisible more button at chevron position.

NOTE: temporary implementation, until design is defined.

nikola-milicevic committed:
Add isReferrerSpam property to StatsTotalRowData, in order to distinguish between spam / nonspam referrers.

nikola-milicevic committed:
Give different text color to parent and referrers marked as spam.

nikola-milicevic committed:
Handle touch events for referrer cell.

nikola-milicevic committed:
Add delegate methods for toggling referrer spam state.

nikola-milicevic committed:
Add action sheets displayed when toggling between referrers spam state.

nikola-milicevic committed:
Update StatsPeriodStore referrers state after successful API call.

nikola-milicevic committed:
Refresh UI when popped, restore receipt listener.

emilylaguna commented on WordPress/Classes/ViewRelated/Stats/Period Stats/SiteStatsPeriodTableViewController.swift:
```diff
@@ -197,6 +209,24 @@ private extension SiteStatsPeriodTableViewController {
+        [toggleSpamAction, cancelAction].forEach {
+            sheet.addAction($0)
+        }
+        present(sheet, animated: true)
```
It looks like this logic is duplicated elsewhere, is there a way we can remove the duplicates?

emilylaguna commented on WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/SiteStatsDetailTableViewController.swift:
```diff
@@ -363,3 +367,24 @@ extension SiteStatsDetailTableViewController: SiteStatsTableHeaderDelegate {
+        [toggleSpamAction, cancelAction].forEach {
+            sheet.addAction($0)
+        }
+        present(sheet, animated: true)
```
It looks like this logic is duplicated elsewhere, is there a way we can remove the duplicates?

emilylaguna commented on WordPress/Classes/ViewRelated/Stats/Shared Views/StatsTotalRow.swift:
```diff
@@ -236,6 +253,23 @@ private extension StatsTotalRow {
+                    moreButton!.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -4),
+                    moreButton!.topAnchor.constraint(equalTo: topAnchor, constant: 4),
+                    moreButton!.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -4),
+                    moreButton!.widthAnchor.constraint(equalTo: moreButton!.heightAnchor)
```
I know this code is temporary but we should avoid using force unwraps.

emilylaguna commented on WordPress/Classes/Stores/StatsPeriodStore.swift:
```diff
@@ -1343,6 +1343,45 @@ extension StatsPeriodStore {
+            return
+        }
+
+        for (i, r) in state.topReferrers!.referrers.enumerated() {
+            if r.children.isEmpty {
+                if r.url?.host == referrerDomain {
+                    let currentValue = state.topReferrers!.referrers[i].isSpam
+                    statsServiceRemote?.toggleSpamState(for: referrerDomain, currentValue: currentValue, success: { [weak self] in
+                        guard let self = self else {
+                            return
+                        }
+                        self.state.topReferrers!.referrers[i].isSpam.toggle()
+                        DDLogInfo("Stats Period: Referrer \(r.title) isSpam set to \(self.state.topReferrers!.referrers[i].isSpam)")
+                    }, failure: { error in
+                        DDLogInfo("Stats Period: Couldn't toggle spam state for referrer \(r.title), reason: \(error.localizedDescription)")
+                    })
+                    break
+                }
+            } else if r.children.first?.url?.host == referrerDomain {
+                let currentValue = state.topReferrers!.referrers[i].isSpam
+                statsServiceRemote?.toggleSpamState(for: referrerDomain, currentValue: currentValue, success: { [weak self] in
+                    guard let self = self else {
+                        return
+                    }
+                    self.state.topReferrers?.referrers[i].isSpam.toggle()
+                    for (j, _) in r.children.enumerated() {
+                        self.state.topReferrers!.referrers[i].children[j].isSpam.toggle()
+                    }
+                    DDLogInfo("Stats Period: Referrer \(r.title) isSpam set to \(self.state.topReferrers!.referrers[i].isSpam)")
+                }, failure: { error in
+                    DDLogInfo("Stats Period: Couldn't toggle spam state for referrer \(r.title), reason: \(error.localizedDescription)")
+                })
+                break
+            }
+        }
```
I left some notes below on how we can improve the readability of this code:

- Variable naming should be more explicit
- There is some duplicated code that can be reduced
- We should avoid using force unwraps
- Error Handling: If the request fails for whatever reason the user may not be aware of that. I think using the `Notice` to let the user know what happened could be helpful
- Success Handling: While the table does change color it can be hard to tell that anything happened. I wonder if we can display a `Notice` for the success as well?

Also, it looks like a lot of this code is just to retrieve a `StatsReferrer` obj from the store and check the state, is there a better way to accomplish this? Maybe pass the original `StatsReferrer` from the row to this method?

What do you think?

nikola-milicevic replied:
I agree with everything :)

emilylaguna commented on WordPress/Classes/Stores/StatsPeriodStore.swift:
```diff
@@ -1343,6 +1343,45 @@ extension StatsPeriodStore {
+                        guard let self = self else {
+                            return
+                        }
+                        self.state.topReferrers!.referrers[i].isSpam.toggle()
```
TIL! I was not aware that Swift has a `toggle` method for Bools!

emilylaguna reviewed (changes requested):
👋 @nikola-milicevic! Nice work so far, I left some comments below, let me know your thoughts. 

I also encountered a few issues with the logic as well:

When comparing the ability to mark an item as spam on wordpress.com and in the PR, the PR allowed me to mark some items as spam when it shouldn't have. 

Some examples:
- Facebook
- Search Engines
- WordPress.com Reader

It may be worth exploring how this was implemented on .com, luckily this code is open source  and lives here [wp-calypso](https://github.com/Automattic/wp-calypso/). 

I'm also thinking we should add some units tests for the store mark as spam logic, what do you think?

nikola-milicevic committed:
Move showSpamActionSheet to UIViewController extension

nikola-milicevic committed:
Make moreButton constant.

nikola-milicevic committed:
Simplify toggleSpamState logic.

nikola-milicevic committed:
Add StatsPeriodStoreDelegate

nikola-milicevic committed:
Display notice when marking / unmarking succeeded / failed.

nikola-milicevic committed:
Add StatsPeriodStore mark / unmark as spam tests.

nikola-milicevic committed:
Discard changes in Localizable.strings file

nikola-milicevic committed:
Add logic for deciding if referrer can be marked as spam or not

nikola-milicevic commented:
@emilylaguna Thanks for the comments, I updated the code, including additional logic to distinguish between referrers that can and cannot be marked as spam, and added tests for store mark as spam logic.

emilylaguna commented on WordPress/Classes/Stores/StatsPeriodStore.swift:
```diff
@@ -172,15 +172,38 @@ struct PeriodStoreState {
 
+protocol StatsPeriodStoreDelegate: AnyObject {
+    func didChangeSpamState(for referrerDomain: String, isSpam: Bool)
+    func chagingSpamStateForReferrerDomainFailed(oldValue: Bool)
```
```suggestion
    func changingSpamStateForReferrerDomainFailed(oldValue: Bool)
```

nikola-milicevic replied:
Oops 😁

emilylaguna commented on WordPress/Classes/Stores/StatsPeriodStore.swift:
```diff
@@ -172,15 +172,38 @@ struct PeriodStoreState {
+
+extension StatsPeriodStoreDelegate where Self: UIViewController {
+    func didChangeSpamState(for referrerDomain: String, isSpam: Bool) {
+        let markedText = NSLocalizedString("Marked as spam", comment: "Indicating that referrer was marked as spam").lowercased()
```
Do we need the `.lowercased()` ?

nikola-milicevic replied:
Well not sure what's the general approach (if we need lowercase to be displayed at all, and not `www.someting.com Marked as spam`). Should localized file contain already lowercase string, or we do the modification at the call site (like i did)?

emilylaguna replied:
I'd say to use the lowercase string directly instead of `lowercased` since I trust the translator to use the correct lowercased string more than I trust swifts interpretation.

emilylaguna commented on WordPress/Classes/ViewRelated/Stats/Shared Views/StatsTotalRow.swift:
```diff
@@ -41,11 +43,17 @@ struct StatsTotalRowData {
     }
+
+    var canMarkReferrerAsSpam: Bool {
+        disclosureURL != nil && disclosureURL!.absoluteString.contains(name) ||
```
Can we remove the force unwrap?

emilylaguna replied:
I think this makes a good candidate for some unit testing, what do you think?

nikola-milicevic replied:
Yes, I agree, I'll add tests for this. Also, I thought force unwrap is okay in this case since with this short circuit evaluation crash can never happen (since first check is `!= nil`). But I can remove the `!` operator.

emilylaguna replied:
>Also, I thought force unwrap is okay in this case since with this short circuit evaluation crash can never happen (since first check is != nil). But I can remove the ! operator

While this is an appropriate use we like to avoid using force unwraps in the codebase.

emilylaguna commented on WordPress/Classes/ViewRelated/Stats/Period Stats/SiteStatsPeriodTableViewController.swift:
```diff
@@ -196,7 +208,12 @@ private extension SiteStatsPeriodTableViewController {
+// MARK: - StatsPeriodStoreDelegate
+
+extension SiteStatsPeriodTableViewController: StatsPeriodStoreDelegate {
+    /* using default implementtion in protocol extension */
```
```suggestion
    /* using default implementation in protocol extension */
```

emilylaguna commented on WordPress/Classes/ViewRelated/Stats/Shared Views/Stats Detail/SiteStatsDetailTableViewController.swift:
```diff
@@ -267,6 +268,12 @@ private extension SiteStatsDetailTableViewController {
+// MARK: - StatsPeriodStoreDelegate
+
+extension SiteStatsDetailTableViewController: StatsPeriodStoreDelegate {
+    /* using default implementtion in protocol extension */
```
```suggestion
    /* using default implementation in protocol extension */
```

emilylaguna commented on WordPress/Classes/Stores/StatsPeriodStore.swift:
```diff
@@ -172,15 +172,38 @@ struct PeriodStoreState {
+}
+
+extension StatsPeriodStoreDelegate where Self: UIViewController {
+    func didChangeSpamState(for referrerDomain: String, isSpam: Bool) {
```
This is a very interesting approach that I haven't seen used before. Nice work.

emilylaguna reviewed (changes requested):
Nice work implementing the logic, it looks to be working nicely now 🙌

I left a few comments below, let me know what you think.

nikola-milicevic committed:
Fix typos.

nikola-milicevic committed:
Change .lowercased() for lowercased localized string.

nikola-milicevic committed:
Remove force unwrap from canMarkReferrerAsSpam.

nikola-milicevic committed:
Add StatsTotalRowDataTests.

nikola-milicevic committed:
Update podfile, use trial branch for WordPressKit

emilylaguna reviewed (approved):
LGTM!
